gtsfm/frontend/cacher/detector_descriptor_cacher.py

```diff
@@ -0,0 +1,89 @@
+
+CACHE_ROOT_PATH = Path(__file__).resolve().parent.parent.parent.parent / "cache"
+
+
+class DetectorDescriptorCacher(DetectorDescriptorBase):
+    """Cacher for detector-descriptor output on disk, keyed on the input."""
+
+    def __init__(self, detector_descriptor_obj: DetectorDescriptorBase) -> None:
+        """Initializes the cacher with the actual detector-descriptor object.
+
+        Args:
+            detector_descriptor_obj: detector-descriptor to use in case of cache miss.
+        """
+        super().__init__(max_keypoints=detector_descriptor_obj.max_keypoints)
+        self._detector_descriptor = detector_descriptor_obj
+        # TODO: make the obj cache key dependent on the code
+        self._detector_descriptor_obj_cache_key = type(self._detector_descriptor).__name__
+
+    def __get_cache_path(self, cache_key: str) -> Path:
+        """Gets the cache path from the cache key."""
+        return CACHE_ROOT_PATH / "detector_descriptor" / "{}.pbz2".format(cache_key)
+
+    def __generate_cache_key(self, image: Image) -> str:
+        """Generates the cache key from the input image and underlying detector descriptor."""
+        input_key = cache_utils.generate_hash_for_image(image)
+        return "{}_{}".format(self._detector_descriptor_obj_cache_key, input_key)
+
+    def __load_result_from_cache(self, image: Image) -> Optional[Tuple[Keypoints, np.ndarray]]:
+        """Load cached result, if they exist."""
+        cache_path = self.__get_cache_path(cache_key=self.__generate_cache_key(image=image))
+        cached_data = io_utils.read_from_compressed_file(cache_path)
+        if cached_data is None:
+            return None
+        return cached_data["keypoints"], cached_data["descriptors"]
+
+    def __save_result_to_cache(self, image: Image, keypoints: Keypoints, descriptors: np.ndarray) -> None:
+        """Save the results to the cache."""
+        cache_path = self.__get_cache_path(cache_key=self.__generate_cache_key(image=image))
+        data = {"keypoints": keypoints, "descriptors": descriptors}
+        io_utils.write_to_compressed_file(data, cache_path)
+
+    def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
+        """Perform feature detection as well as their description, with caching.
+
+        If the results are in the cache, they are fetched and returned. Otherwise, the `detect_and_describe` of the
+        underlying object's API is called and the results are cached.
+
+        Refer to detect() in DetectorBase and describe() in DescriptorBase for
+        details about the output format.
+
+        Args:
+            image: the input image.
+
+        Returns:
+            Detected keypoints, with length N <= max_keypoints.
+            Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
+        """
+        cached_data = self.__load_result_from_cache(image)
+
+        if cached_data is not None:
+            return cached_data
```

Comment:
Should we have a print statement here to let the user know that cached results are being returned for the keypoints and descriptors?

Reply:
Umm it will lead to lot of lines in the logs. Ideally we would want aggregate values to be printed, but we do not have a framework to record such metrics right now. 

If we do want to have this appear onto stdout, I'd prefer using DEBUG log level and configuring the CI to have INFO and above log level.

Reply:
Nope. Cannot do it selectively. It will either be for all or for none of them.